### PR TITLE
JAMES-3435 CRDT friendly version of CassandraACLDAO 

### PR DESCRIPTION
**Pull request by chibenwa**
On a real plateform moves the breaking point from 5400 users to ~10000 users (breaking point being p99 of 10s + 1% of error)

Very good latencies are observed until 9000 users.

Handled load: 600 req/s instead of 325 req/s.

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAO.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    Mono<ACLDiff> doUpdateAcl(CassandraId cassandraId, Function<ACLWithVersion, ACLWithVersion> aclTransformation, MailboxACL replacement) {
+        return getAclWithVersion(cassandraId)
+            .flatMap(aclWithVersion ->
+                    updateStoredACL(cassandraId, aclTransformation.apply(aclWithVersion))
+                            .map(newACL -> ACLDiff.computeDiff(aclWithVersion.mailboxACL, newACL)))
+            .switchIfEmpty(insertACL(cassandraId, replacement)
+                    .map(newACL -> ACLDiff.computeDiff(MailboxACL.EMPTY, newACL)))
+            .single()
+            .retry(maxAclRetry);
```
Comment (LanKhuat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAO.java]: Indentation seems weird?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraAclModule.java [hunk 2]
```diff
@@ -42,6 +44,17 @@
             .addPartitionKey(CassandraACLTable.ID, timeuuid())
             .addColumn(CassandraACLTable.ACL, text())
             .addColumn(CassandraACLTable.VERSION, bigint()))
+
+        .table(CassandraACLV2Table.TABLE_NAME)
+        .comment("Holds mailbox ACLs. This table do not rely on a JSON representation nor on LWT, contrary to the acl table it replaces.")
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraAclModule.java:49]: could you describe what it does (positive description, not description in opposition to previous one) ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAOV2.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    private PreparedStatement prepareDelete(Session session) {
+        return session.prepare(
+            QueryBuilder.delete().from(CassandraACLV2Table.TABLE_NAME)
+                .where(eq(CassandraACLV2Table.ID, bindMarker(CassandraACLV2Table.ID))));
+    }
+
+    private PreparedStatement prepareInsertRights(Session session) {
+        return session.prepare(
+            update(CassandraACLV2Table.TABLE_NAME)
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAOV2.java:78]: Both are the same!

**Review by mbaechler — COMMENTED** [review]
You talk about CRDT but I can't find which CRDT structure you use, could you elaborate?
Also, I lack an overview of the solution you implemented.
What I see it that you store individual right entries and mutate them over time: is it the actual difference ? Do I miss something else ? Why did we use LWT in the first place ?

Aside these missing information, It looks like a good idea.

**Comment by chibenwa** [comment]
Thanks @mbaechler for the look. I miss time to do a throughout answer now but the idea is to be conflict free on the table of truth.

We would benefit from "applicative read repairs" and "corrective tasks" on ACL as well but they might be coming later.

> Aside these missing information, It looks like a good idea.

Next step should be to act that we have no condstore on top of Cassandra and relax LWT there too. Though it likely is less critical.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAOV2.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAOV2.java:78]: Still for the sake of syntax... let me fix quickly

[the remark this answers is not in the document]
Comment (Arsnael) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLDAOV2.java:78]: Can't find the equivalent qui an insert... I'm ok with it

**Comment by Arsnael** [comment]
```
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraGetMessageListMethodTest>GetMessageListMethodTest.getMessageListShouldExcludeMessagesWhenInMailboxesFilterMatches:1000 1 expectation failed.
[691a51b3ff19f90638294d94c23fa832b3234b5c] Expected status code <200> but was <500>.
[691a51b3ff19f90638294d94c23fa832b3234b5c] 
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraGetMessageListMethodTest>GetMessageListMethodTest.getMessageListShouldFetchUnreadMessagesInMailbox:661 1 expectation failed.
[691a51b3ff19f90638294d94c23fa832b3234b5c] Expected status code <200> but was <500>.
[691a51b3ff19f90638294d94c23fa832b3234b5c] 
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraGetMessageListMethodTest>GetMessageListMethodTest.getMessageListShouldNotExcludeMessagesWhenInMailboxesFilterMatchesMailboxAndText:1025 1 expectation failed.
[691a51b3ff19f90638294d94c23fa832b3234b5c] Expected status code <200> but was <500>.
[691a51b3ff19f90638294d94c23fa832b3234b5c] 
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMailboxesMethodCucumberTest 
[691a51b3ff19f90638294d94c23fa832b3234b5c] Expecting:
[691a51b3ff19f90638294d94c23fa832b3234b5c]  <500>
[691a51b3ff19f90638294d94c23fa832b3234b5c] to be equal to:
[691a51b3ff19f90638294d94c23fa832b3234b5c]  <200>
[691a51b3ff19f90638294d94c23fa832b3234b5c] but was not.
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMailboxesMethodCucumberTest 
[691a51b3ff19f90638294d94c23fa832b3234b5c] Expecting:
[691a51b3ff19f90638294d94c23fa832b3234b5c]  <500>
[691a51b3ff19f90638294d94c23fa832b3234b5c] to be equal to:
[691a51b3ff19f90638294d94c23fa832b3234b5c]  <200>
[691a51b3ff19f90638294d94c23fa832b3234b5c] but was not.
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR] Errors: 
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraFilterTest>FilterTest.messageShouldBeAppendedInInboxWhenFromDoesNotMatchRule:1213 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraFilterTest>FilterTest.messageShouldBeAppendedInSpecificMailboxWhenFromRuleMatches:879 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.attachmentsShouldBeRetrievedWhenChainingSetMessagesAndGetMessagesTextAttachment:4518 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.sendingAMailShouldLeadToAppropriateMailboxCountersOnOutbox:1114 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.sendingAMailShouldLeadToAppropriateMailboxCountersOnSent:1233 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldMoveMessageInSentWhenMessageIsSent:2632 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldUpdateIsAnsweredWhenInReplyToHeaderSentViaOutbox:5027->SetMessagesMethodTest.receiveFirstMessage:5277 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldUpdateIsForwardedWhenXForwardedHeaderSentViaOutbox:5074->SetMessagesMethodTest.receiveFirstMessage:5277 » ConditionTimeout
[691a51b3ff19f90638294d94c23fa832b3234b5c] [ERROR]   CassandraVacationIntegrationTest>VacationIntegrationTest.jmapVacationShouldGenerateAReplyWhenActive:127 » ConditionTimeout
```

**Comment by Arsnael** [comment]
```
[8f28d3c92193fbd3ba4d823266da5d5e4bc72358] [ERROR] Errors: 
[8f28d3c92193fbd3ba4d823266da5d5e4bc72358] [ERROR]   CassandraMessageIdManagerStorageTest.setFlagsShouldUpdateFlagsWhenWriteDelegated » Mailbox
```

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[97f982e5508e6026990cf4ee98866f4c7fc9c1b3] [ERROR]   FixingGhostMailboxTest.webadminCanMergeTwoMailboxes:259 1 expectation failed.
[97f982e5508e6026990cf4ee98866f4c7fc9c1b3] JSON path [0][1].messageIds doesn't match.
[97f982e5508e6026990cf4ee98866f4c7fc9c1b3] Expected: a collection with size <2>
[97f982e5508e6026990cf4ee98866f4c7fc9c1b3]   Actual: [6f6fdea0-34cb-11eb-985a-fd4ddeca3446]
```

test this please

**Comment by chibenwa** [comment]
`DiscardProtocolTest.testRecord:84 »  Timed out wait for be notified that read ...`

**Comment by chibenwa** [comment]
This PR is cursed! Bad code defends itself! :-)

`docker: Error response from daemon: No such container: git-publish-acl-v2-238c84d8-61f9-4a7e-a15d-6b397d436c94.`

Unrelated

test this please

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
> You talk about CRDT but I can't find which CRDT structure you use, could you elaborate?

Each rights is a set, thus conflicts in the rights of each entries can be resolved through the set crdt data type built in cassandra.

Collections can't be nested.

So I opted for a clustering key per entry. Adding/removing entries is thus independent.

The flip side is the set/reset. Each entry (clustering key) needs to be individually inserted. I tried to bring in some consistency using batch statements (all statements are eventually applied).

**Comment by chibenwa** [comment]
Thonking about it: we can have event sourcing here.

Decisions would stay strongly consistent while offering non LWT on reads (using the exisying DAOs.

I think an ADR is needed here too.

**Comment by chibenwa** [comment]
https://github.com/apache/james-project/pull/271 `[ADR] Against the use of Lightweight Transaction in Cassandra code base` provides context around this work.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java [hunk 4]
```diff
@@ -19,66 +19,169 @@
...
+
+        @Override
+        public Mono<ACLDiff> setACL(CassandraId cassandraId, MailboxACL mailboxACL) {
+            return Mono.from(eventSourcingSystem.dispatch(new SetACLCommand(new MailboxAggregateId(cassandraId), mailboxACL)))
+                .flatMapIterable(events -> events)
+                .filter(ACLReseted.class::isInstance)
+                .map(ACLReseted.class::cast)
+                .map(ACLReseted::getAclDiff)
+                .next()
+                .switchIfEmpty(Mono.defer(() -> Mono.error(new MailboxException("Unable to update ACL"))));
```
Comment (Arsnael) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java:140]: Maybe the log error should be `"Unable to update ACL"`, to differentiate with the one above?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java [hunk 5]
```diff
@@ -78,6 +84,7 @@
     private static final Mailbox MAILBOX_BIS = new Mailbox(MAILBOX_PATH, UID_VALIDITY, MAILBOX_ID_2);
 
     private static final CassandraModule MODULES = CassandraModule.aggregateModules(
+        CassandraEventStoreModule.MODULE(),
         CassandraMailboxModule.MODULE,
         CassandraSchemaVersionModule.MODULE,
         CassandraAclModule.MODULE);
```
Comment (Arsnael) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java:90]: can you reorder this one while we are at it?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/MailboxAggregateModule.java [hunk 6]
```diff
@@ -39,6 +40,7 @@
 
 public interface MailboxAggregateModule {
     CassandraModule MODULE = CassandraModule.aggregateModules(
+            CassandraEventStoreModule.MODULE(),
             CassandraSchemaVersionModule.MODULE,
             CassandraAclModule.MODULE,
```
Comment (Arsnael) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/MailboxAggregateModule.java:45]: idem

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/AclV2MigrationTest.java [hunk 7]
```diff
@@ -54,7 +62,8 @@
     public static final CassandraModule MODULES = CassandraModule.aggregateModules(
         CassandraMailboxModule.MODULE,
         CassandraAclModule.MODULE,
-        CassandraSchemaVersionModule.MODULE);
+        CassandraSchemaVersionModule.MODULE,
+        CassandraEventStoreModule.MODULE());
```
Comment (Arsnael) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/AclV2MigrationTest.java:66]: idem

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java [hunk 8]
```diff
@@ -100,28 +107,32 @@ void setUp() {
...
-            new CassandraACLDAOV1(cassandra.getConf(),
-                CassandraConfiguration.DEFAULT_CONFIGURATION,
-                cassandraCluster.getCassandraConsistenciesConfiguration()),
-            new CassandraACLDAOV2(cassandra.getConf()),
-            new CassandraSchemaVersionManager(new CassandraSchemaVersionDAO(cassandra.getConf())));
-        versionDAO = new CassandraSchemaVersionDAO(cassandra.getConf());
 
+        versionDAO = new CassandraSchemaVersionDAO(cassandra.getConf());
         versionDAO.truncateVersion()
             .then(versionDAO.updateVersion(new SchemaVersion(7)))
```
Comment (Arsnael) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java:113]: Why wouldn't we want to test it with the new ACL Store?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java:90]: :-(

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/MailboxAggregateModule.java (hunk 6 above)
Comment (chibenwa) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/MailboxAggregateModule.java:45]: That's cosmetic

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java (hunk 8 above)
Comment (Arsnael) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java:113]: Is it an issue? I see a lot of disabled tests for this against the new ACL store

Comment (chibenwa) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java:113]: The sames as against the old one.

**Comment by chibenwa** [comment]
```
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] Apache James :: Server :: Web Admin server integration tests :: Memory FAILURE [20:05 min]
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] ------------------------------------------------------------------------
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] BUILD FAILURE
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] ------------------------------------------------------------------------
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] Total time:  02:49 h
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] Finished at: 2020-12-04T13:14:14Z
[a96d633433fc797496f649514954ffd52ad136f1] [INFO] ------------------------------------------------------------------------
[a96d633433fc797496f649514954ffd52ad136f1] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project memory-webadmin-integration-test: There was a timeout or other error in the fork -> [Help 1]
```

Unrelated

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[566447c6b5f41d74e2c92e6ebfbb1d5ce9ec297b] [ERROR] Failed to execute goal on project james-json: Could not resolve dependencies for project org.apache.james:james-json:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.fasterxml.jackson.datatype:jackson-datatype-guava:jar:2.9.9 -> com.google.guava:guava:jar:25.1-jre -> org.checkerframework:checker-qual:jar:2.0.0: Failed to read artifact descriptor for org.checkerframework:checker-qual:jar:2.0.0: Could not transfer artifact org.checkerframework:checker-qual:pom:2.0.0 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.pom.part (No such file or directory) -> [Help 1]
```

test this please

**Review by LanKhuat — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**